README
verticalize is transforming any terminal output vertically, i.e. the horizontal text is translated vertically.


NEWS
----------------------------------------------------------------------------
transpose 0.1

* First version of this program


src/transpose.c
/*  transpose: outputs horizontal table vertically
    Copyright (C) 2018  Guillaume Koehl <opengk@free.fr>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>. */

/* Written by Guillaume Koehl */

/* Thanks a lot to the K&R Learning C book which
 * almost explicits all the code  and to the GNU
 * coreutils source code which is a great source
 * of inspiration for coding in the linux system*/

#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#define PROG_NAME "transpose"
#define BUFSIZE 8192
#define STREQ(a, b) (*(a) == *(b) && strcmp((a), (b)) == 0)

static struct option const longopts[] = {
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'v'},
};

void usage(int status)
{
    if (status != EXIT_SUCCESS) {
        fprintf(stderr, "Try '%s -h or --help' for more information.\n",
                PROG_NAME);
    } else {
        printf(
            "Usage: %s [FILE]\n\
Display text vertically, for example create a vertical histogram from an"
"horizontal input.\n\n\
Without FILE, read standard input.\n\n\
  -h, --help     Show help and exit\n\
  -v, --version  Show version\n",
            PROG_NAME);
    }
    exit(status);
}

int count_lines(char *input)
{
    int lines = 0;
    for (size_t i = 0; i < strlen(input); ++i)
        if (input[i] == '\n')
            lines++;

    return lines;
}

int count_longest_line(char *input)
{
    int curr_char = 0, max_char = 0, curr_line = 0;
    for (size_t i = 0; i < strlen(input); ++i) {
        curr_char++;
        if (input[i] == '\n') {
            if (curr_char > max_char)
                max_char = curr_char;
            curr_line++;
            curr_char = 0;
        }
    }
    return max_char;
}

void transpose(const char *file)
{
    /* Open Stdin or a File */
    int fd;
    if (STREQ(file, "stdin")) {
        fd = STDIN_FILENO; /* stdin */
    } else {
        fd = open(file, O_RDONLY); /* read from file */
        if (fd == -1)
            perror("Could not open file");
    }

    /* Read file */
    char buf[BUFSIZE];   /* buffer */
    char input[BUFSIZE]; /* all text */
    for (int i = 0; i < BUFSIZE; ++i) {
        input[i] = '\0';
    }

    int read_bytes = 0, tot_bytes = 0; /* bytes */
    while (1) {
        read_bytes = read(fd, buf, BUFSIZE); /* copy to buffer */

        if (read_bytes == 0) /* EOF */
            break;

        if (read_bytes == -1) {
            perror("read");
            break;
        }

        tot_bytes += read_bytes;

        for (int i = 0; i < read_bytes; i++) /* record each line */
            input[i + tot_bytes - read_bytes] = buf[i];
    }

    /* Count lines and longest line chars number */
    int lines = count_lines(input);
    int max_char = count_longest_line(input);

    /* Store data in an array to facilitate transposing */
    int text[1000][2048];
    for (int i = 0; i < lines; ++i)
        for (int j = 0; j < max_char; ++j)
            text[i][j] = 0;

    int curr_char = 0, curr_line = 0;
    for (size_t i = 0; i < strlen(input); ++i) {
        if (input[i] == '\n') {
            curr_char = 0;
            curr_line++;
        } else {
            if (input[i] == '\t') /* tab is converted to space character */
                text[curr_line][curr_char] = ' ';
            else
                text[curr_line][curr_char] = input[i];
            curr_char++;
        }
    }

    /* Transpose the table */
    for (int j = max_char - 2; j >= 0; --j) {
        for (int i = 0; i < lines; ++i) {
            if (text[i][j] != 0)
                printf("%c ", text[i][j]);
            else
                printf("  ");
        }
        printf("\n");
    }
}

int main(int argc, char *argv[])
{
    /* Parameters with getopts */
    int c;
    while ((c = getopt_long(argc, argv, "hv", longopts, NULL)) != -1) {
        switch (c) {
            case 'h': /* help */
                usage(EXIT_SUCCESS);
                break;

            case 'v': /* version */
                printf(
                    "transpose  Copyright (C) 2018  Guillaume Koehl\n\
GNU General Public License v3 (GPL-3). This program comes with ABSOLUTELY NO "
"WARRANTY.\n\
This is free software, and you are welcome to redistribute it under certain "
"conditions.\n");
                exit(EXIT_SUCCESS);
                break;

            default: /* usage */
                usage(EXIT_FAILURE);
        }
    }

    /* Get source and transpose */
    int files = argc - optind;
    if (files == 0)
        transpose("stdin");
    else if (files == 1)
        transpose(argv[1]);
    else
        usage(EXIT_SUCCESS);

    exit(EXIT_SUCCESS);
}


.ycm_extra_conf.py
# This file is NOT licensed under the GPLv3, which is the license for the rest
# of YouCompleteMe.
#
# Here's the license text for this file:
#
# This is free and unencumbered software released into the public domain.
#
# Anyone is free to copy, modify, publish, use, compile, sell, or
# distribute this software, either in source code form or as a compiled
# binary, for any purpose, commercial or non-commercial, and by any
# means.
#
# In jurisdictions that recognize copyright laws, the author or authors
# of this software dedicate any and all copyright interest in the
# software to the public domain. We make this dedication for the benefit
# of the public at large and to the detriment of our heirs and
# successors. We intend this dedication to be an overt act of
# relinquishment in perpetuity of all present and future rights to this
# software under copyright law.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.
#
# For more information, please refer to <http://unlicense.org/>

from distutils.sysconfig import get_python_inc
import platform
import os
import ycm_core

# These are the compilation flags that will be used in case there's no
# compilation database set (by default, one is not set).
# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
flags = [
'-Wall',
'-Wextra',
'-Werror',
'-Wno-long-long',
'-Wno-variadic-macros',
'-fexceptions',
'-DNDEBUG',
# You 100% do NOT need -DUSE_CLANG_COMPLETER and/or -DYCM_EXPORT in your flags;
# only the YCM source code needs it.
'-DUSE_CLANG_COMPLETER',
'-DYCM_EXPORT=',
# THIS IS IMPORTANT! Without the '-x' flag, Clang won't know which language to
# use when compiling headers. So it will guess. Badly. So C++ headers will be
# compiled as C headers. You don't want that so ALWAYS specify the '-x' flag.
# For a C project, you would set this to 'c' instead of 'c++'.
'-x',
'c++',
'-isystem',
'../BoostParts',
'-isystem',
get_python_inc(),
'-isystem',
'../llvm/include',
'-isystem',
'../llvm/tools/clang/include',
'-I',
'.',
'-I',
'./ClangCompleter',
'-isystem',
'./tests/gmock/gtest',
'-isystem',
'./tests/gmock/gtest/include',
'-isystem',
'./tests/gmock',
'-isystem',
'./tests/gmock/include',
'-isystem',
'./benchmarks/benchmark/include',
]

# Clang automatically sets the '-std=' flag to 'c++14' for MSVC 2015 or later,
# which is required for compiling the standard library, and to 'c++11' for older
# versions.
if platform.system() != 'Windows':
  flags.append( '-std=c++11' )


# Set this to the absolute path to the folder (NOT the file!) containing the
# compile_commands.json file to use that instead of 'flags'. See here for
# more details: http://clang.llvm.org/docs/JSONCompilationDatabase.html
#
# You can get CMake to generate this file for you by adding:
#   set( CMAKE_EXPORT_COMPILE_COMMANDS 1 )
# to your CMakeLists.txt file.
#
# Most projects will NOT need to set this to anything; you can just change the
# 'flags' list of compilation flags. Notice that YCM itself uses that approach.
compilation_database_folder = ''

if os.path.exists( compilation_database_folder ):
  database = ycm_core.CompilationDatabase( compilation_database_folder )
else:
  database = None

SOURCE_EXTENSIONS = [ '.cpp', '.cxx', '.cc', '.c', '.m', '.mm' ]

def DirectoryOfThisScript():
  return os.path.dirname( os.path.abspath( __file__ ) )


def IsHeaderFile( filename ):
  extension = os.path.splitext( filename )[ 1 ]
  return extension in [ '.h', '.hxx', '.hpp', '.hh' ]


def FindCorrespondingSourceFile( filename ):
  if IsHeaderFile( filename ):
    basename = os.path.splitext( filename )[ 0 ]
    for extension in SOURCE_EXTENSIONS:
      replacement_file = basename + extension
      if os.path.exists( replacement_file ):
        return replacement_file
  return filename


def FlagsForFile( filename, **kwargs ):
  # If the file is a header, try to find the corresponding source file and
  # retrieve its flags from the compilation database if using one. This is
  # necessary since compilation databases don't have entries for header files.
  # In addition, use this source file as the translation unit. This makes it
  # possible to jump from a declaration in the header file to its definition in
  # the corresponding source file.
  filename = FindCorrespondingSourceFile( filename )

  if not database:
    return {
      'flags': flags,
      'include_paths_relative_to_dir': DirectoryOfThisScript(),
      'override_filename': filename
    }

  compilation_info = database.GetCompilationInfoForFile( filename )
  if not compilation_info.compiler_flags_:
    return None

  # Bear in mind that compilation_info.compiler_flags_ does NOT return a
  # python list, but a "list-like" StringVec object.
  final_flags = list( compilation_info.compiler_flags_ )

  # NOTE: This is just for YouCompleteMe; it's highly likely that your project
  # does NOT need to remove the stdlib flag. DO NOT USE THIS IN YOUR
  # ycm_extra_conf IF YOU'RE NOT 100% SURE YOU NEED IT.
  try:
    final_flags.remove( '-stdlib=libc++' )
  except ValueError:
    pass

  return {
    'flags': final_flags,
    'include_paths_relative_to_dir': compilation_info.compiler_working_dir_,
    'override_filename': filename
}


src/verticalize.c
/*  verticalize: outputs horizontal table vertically
    Copyright (C) 2018  Guillaume Koehl <opengk@free.fr>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>. */

/* Written by Guillaume Koehl */

/* Thanks a lot to the K&R Learning C book which
 * almost explicits all the code  and to the GNU
 * coreutils source code which is a great source
 * of inspiration for coding in the linux system*/

#include <fcntl.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#define PROG_NAME "verticalize"
#define BUFSIZE 8192
#define STREQ(a, b) (*(a) == *(b) && strcmp((a), (b)) == 0)

void usage(int status)
{
    if (status != EXIT_SUCCESS) {
        fprintf(stderr, "Try '%s -h or --help' for more information.\n",
                PROG_NAME);
    } else {
        printf(
            "Usage: %s [FILE]\n\
Display text vertically, for example create a vertical histogram from an"
"horizontal input.\n\n\
Without FILE, read standard input.\n\n\
  -b, --bottom-start Verticalize starting from bottom\n\
  -h, --help         Show help and exit\n\
  -v, --version      Show version\n",
            PROG_NAME);
    }
    exit(status);
}

int count_lines(char *input)
{
    int lines = 0;
    for (size_t i = 0; i < strlen(input); ++i)
        if (input[i] == '\n')
            lines++;

    return lines;
}

int count_longest_line(char *input)
{
    int curr_char = 0, max_char = 0, curr_line = 0;
    for (size_t i = 0; i < strlen(input); ++i) {
        curr_char++;
        if (input[i] == '\n') {
            if (curr_char > max_char)
                max_char = curr_char;
            curr_line++;
            curr_char = 0;
        }
    }
    return max_char;
}

void verticalize(const char *file, bool bottom_start)
{
    /* Open Stdin or a File */
    int fd;
    if (STREQ(file, "stdin")) {
        fd = STDIN_FILENO; /* stdin */
    } else {
        fd = open(file, O_RDONLY); /* read from file */
        if (fd == -1)
            perror("Could not open file");
    }

    /* Read file */
    char buf[BUFSIZE];   /* buffer */
    char input[BUFSIZE]; /* all text */
    for (int i = 0; i < BUFSIZE; ++i) {
        input[i] = '\0';
    }

    int read_bytes = 0, tot_bytes = 0; /* bytes */
    while (1) {
        read_bytes = read(fd, buf, BUFSIZE); /* copy to buffer */

        if (read_bytes == 0) /* EOF */
            break;

        if (read_bytes == -1) {
            perror("read");
            break;
        }

        tot_bytes += read_bytes;

        for (int i = 0; i < read_bytes; i++) /* record each line */
            input[i + tot_bytes - read_bytes] = buf[i];
    }

    /* Count lines and longest line chars number */
    int lines = count_lines(input);
    int max_char = count_longest_line(input);

    /* Store data in an array to facilitate transposing */
    int text[1000][2048];
    for (int i = 0; i < lines; ++i)
        for (int j = 0; j < max_char; ++j)
            text[i][j] = 0; /* initialize table */

    int curr_char = 0, curr_line = 0;
    for (size_t i = 0; i < strlen(input); ++i) { /* fill table */
        if (input[i] == '\n') {
            curr_char = 0;
            curr_line++;
        } else {
            if (input[i] == '\t') /* tab converted to space character */
                text[curr_line][curr_char] = ' ';
            else
                text[curr_line][curr_char] = input[i];
            curr_char++;
        }
    }

    /* Verticalize */
    if (bottom_start == false) {
        /* Verticalize from top */
        for (int j = 0; j <= max_char - 2; ++j) {
            for (int i = 0; i < lines; ++i) {
                if (text[i][j] != 0)
                    printf("%c ", text[i][j]);
                else
                    printf("  ");
            }
            printf("\n");
        }
    } else {
        /* Verticalize from bottom */
        for (int j = max_char - 2; j >= 0; --j) {
            for (int i = 0; i < lines; ++i) {
                if (text[i][j] != 0)
                    printf("%c ", text[i][j]);
                else
                    printf("  ");
            }
            printf("\n");
        }
    }
}

int main(int argc, char *argv[])
{
    /* Variables set depending on options */
    bool bottom_start = false;

    static struct option const longopts[] = {
        {"bottom", no_argument, NULL, 'b'},
        {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, 'v'},
        {NULL, 0, NULL, 0}
    };

    /* Parameters with getopts */
    int c;
    while ((c = getopt_long(argc, argv, "bhv", longopts, NULL)) != -1) {
        switch (c) {
            case 'b': /* verticalize from bottom */
                bottom_start = true;
                break;
            case 'h': /* help */
                usage(EXIT_SUCCESS);
                break;
            case 'v': /* version */
                printf(
                    "verticalize  Copyright (C) 2018  Guillaume Koehl\n\
GNU General Public License v3 (GPL-3). This program comes with ABSOLUTELY NO "
"WARRANTY.\n\
This is free software, and you are welcome to redistribute it under certain "
"conditions.\n");
                exit(EXIT_SUCCESS);
                break;
            
            default: /* usage */
                usage(EXIT_FAILURE);
        }
    }

    /* Get source and verticalize */
    int files = argc - optind;
    if (files == 0)
        verticalize("stdin", bottom_start);
    else if (files == 1)
        verticalize(argv[optind], bottom_start);
    else
        usage(EXIT_SUCCESS);

    exit(EXIT_SUCCESS);
}
